### DISCUSSION TITLE
Production deployment: Can I use a shared static cache for multiple nodes?

### DISCUSSION THREAD

**Discussion by janus-reith** [vercel/next.js#11428] [Help]
I‘m building a nextjs application which heavily used fallback rendering and the revalidate parameter to serve a big product catalog.

I wonder what the best way would be for production deployments to handle this efficiently besides using Zeit Now. (Probably will use that anyways, but want to know that there are alternatives).

If I load balance multiple nextjs instances, each of them would rebuild the fallback pages/revalidate pages on their own which seems really inefficient for a large catalog.
Would it be viable to let all of these instances share the same .next cache folder, or would that lead to issues if many of them access it simultaneously?
Also, as currently these multiple nextjs instances are just docker containers on the same host, that would work easily, but he same might not be the case if I have them on entirely separate hosts.

Is load balancing multiple instances actually even a good approach? Most of the time I just want to serve static prerendered content, and would only need more computing power for the api routes and maybe for rendering the fallback or to–be–revalidated pages if many are requested on the same time.

For a serverless deployment I guess things will be easier as most of the time content could be served from one static source, and new functions would be called to render a page one time if certain criteria are met(doesn’t exist yet/need to revalidate) but serverless nextjs doesn’t support the new static functions yet.

All in all I‘m struggling to find a proper deployment example for this and wonder if there might be a more clever way than simply having several containers running nextjs and balance between them.

**Comment by janus-reith** [comment]
After observing a bit how the static pages in .nextjs are handles, I simply created a docker volume for `.next/server/static` which is shared between the containers - If I did'nt miss anything, nextjs just checks for their presence on the filesystem and isn't managing a separate list of cached pages so it should work out. (Could someone confirm?)
This seems to work without issues so far.

**Reply by janus-reith** [reply]
Sadly this results in issues when the build id changes between deployments.
It seem like It won’t work with an emty folder, I wonder if I could just add a startup step where I delete existing deployments and copy over files from static to server/static ?

What are your deployment strategies?

**Reply by jnv** [reply]
Hi, just to chime in regarding the build ID, you can stabilize that across instances by using e.g. commit ID. See the [docs](https://nextjs.org/docs/api-reference/next.config.js/configuring-the-build-id). There's also a package to help precisely with that: https://github.com/nexdrew/next-build-id

**Reply by janus-reith** [reply]
@jnv Thanks for the link!
In my case, all started containers would use the same image and already have the same build id.
The issue I described would most likely still occur when using the id of the last commit.
It has to do with the way docker volumes are initialized: After the build is finished, there will be data in `.next/server/static`. If not present yet, docker will take over existing content from the image's `.next/server/static` path into the volume.
For subsequent starts, it wouldn't do that again, resulting in the correct folder with build id for the current build missing when a new build is deployed. Therefore, each  time a new image is deployed, all nextjs containers would need to stop so the volume can be removed, so it can be recreated with the new data again, which certainly is not a nice solution.

A custom little script executing during container build can make sure these files are stored outside that path, while on container startup another script could copy them from that destination to `.next/server/static`, plus maybe purge data from old build ids aswell.
But I'm not sure what the most streamlined solution would look like here.

**Comment by janus-reith** [comment]
Related: https://github.com/zeit/next.js/discussions/11552#discussioncomment-2804

**Reply by alfrescian** [reply]
Just stumpled upon this discussion while writing #29291.
Looks to me, that with current Next version (11.1.2) ISR/SSG output is no more stored in '.next/server/static' but in `.next/server/pages`.
Thus you don't have any build-id  or hash in the path & I'm wondering how to use a shared storage for ISR pages when using more sophisticated deployment feature like swapping slots.

**Comment by LuizHAP** [comment]
Vercel has created a new update (12.1.0) and how it should work in the multi-node scenario after this update (https://github.com/vercel/next.js/discussions/11552). Will it update all nodes, or just the last one? Or depends on the case?
